## Problem 18: Maximum Path Sum I

By starting at the top of the triangle below and moving to adjacent numbers on the row below, the maximum total from top to bottom is 23.

\begin{align}
    &{\color{red}3}\\
    {\color{red}7}&\ 4\\
    2\ & {\color{red}4}\ 6\\
    8\ 5 &\ {\color{red}9}\ 3
\end{align}

That is, $3 + 7 + 4 + 9 = 23$.

Find the maximum total from top to bottom of the triangle below:

\begin{align}
    &75\\
    95 &\ 64\\
    17\ &47\ 82\\
    18\ 35 &\ 87\ 10\\
    20\ 04\ &82\ 47\ 65\\
    19\ 01\ 23 &\ 75\ 03\ 34\\
    88\ 02\ 77\ &73\ 07\ 63\ 67\\
    99\ 65\ 04\ 28 &\ 06\ 16\ 70\ 92\\
    41\ 41\ 26\ 56\ &83\ 40\ 80\ 70\ 33\\
    41\ 48\ 72\ 33\ 47 &\ 32\ 37\ 16\ 94\ 29\\
    53\ 71\ 44\ 65\ 25\ &43\ 91\ 52\ 97\ 51\ 14\\
    70\ 11\ 33\ 28\ 77\ 73 &\ 17\ 78\ 39\ 68\ 17\ 57\\
    91\ 71\ 52\ 38\ 17\ 14\ &91\ 43\ 58\ 50\ 27\ 29\ 48\\
    63\ 66\ 04\ 68\ 89\ 53\ 67 &\ 30\ 73\ 16\ 69\ 87\ 40\ 31\\
    04\ 62\ 98\ 27\ 23\ 09\ 70\ &98\ 73\ 93\ 38\ 53\ 60\ 04\ 23\\
\end{align}

*NOTE*: As there are only 16384 routes, it is possible to solve this problem by trying every route. However, Problem 67, is the same challenge with a triangle containing one-hundred rows; it cannot be solved by brute force, and requires a clever method!

In [1]:
# Search up the tree backwards, creating a new tree as we go.
# When two branches meet, calculate the highest one and insert value of total route into the space of the node.
# Have a thrid tree where the nodes are the x co-ordinate of the branches chosen.
#
# eg. 
#    3                3
#   7 4              7 4
#  2 4 6            2 4 6
# 8 5 9 3          8 5 9 3
#
#     3                   3
#    7 4                7   4
#  10 13 12          [0] [2] [2]
# x  x  x  x        x   x   x   x
#
#      3                    3
#    20 17             [1,0] [1,2]
#   x  x  x           x     x     x 
# x  x  x  x       x     x    x     x
#
#      23                 [0,1,0]
#     x  x               x     x
#   x  x  x           x     x     x 
# x  x  x  x       x     x    x     x

In [2]:
# Tree of values.
triangle = [
    [75],
    [95,64],
    [17,47,82],
    [18,35,87,10],
    [20, 4,82,47,65],
    [19, 1,23,75, 3,34],
    [88, 2,77,73, 7,63,67],
    [99,65, 4,28, 6,16,70,92],
    [41,41,26,56,83,40,80,70,33],
    [41,48,72,33,47,32,37,16,94,29],
    [53,71,44,65,25,43,91,52,97,51,14],
    [70,11,33,28,77,73,17,78,39,68,17,57],
    [91,71,52,38,17,14,91,43,58,50,27,29,48],
    [63,66, 4,68,89,53,67,30,73,16,69,87,40,31],
    [ 4,62,98,27,23, 9,70,98,73,93,38,53,60, 4,23],
]

# Tree of highest values. Prefill the values with the values of the tree.
highest_value_triangle = []
for y in range(len(triangle)):
    highest_value_triangle.append([])
    for x in range(len(triangle[y])):
        highest_value_triangle[y].append(triangle[y][x])

# Tree of coordinates. Prefill with x coordinates.
coordinate_triangle = []
for y in range(len(triangle)):
    coordinate_triangle.append([])
    for x in range(len(triangle[y])):
        coordinate_triangle[y].append([x])

In [3]:
''' 
Finds highest branch for one node. Returns the value and co-ordinate.

eg:
>>> best_branch(2, 3)
>>> (82, 2)
'''

def best_branch(node_x_coordinate : int, node_y_coordinate : int) -> list:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Finds highest branch for one node. Returns the value and co-ordinate.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : int : node_x_coordinate : The x co-ordinate of the node.
        : int : node_y_coordinate : The y co-ordinate of the node.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : list : The following information is returned:
            : int : The value of the highest branch.
            : int : The x co-ordinate of the highest branch.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> best_branch(2, 3)
        >>> (82, 2)
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(node_x_coordinate, int): raise ValueError('Please enter an integer >= 0 for the node_x_coordinate argument.')
    if node_x_coordinate < 0: raise ValueError('Please enter an integer >= 0 for the node_x_coordinate argument.')
    if not isinstance(node_y_coordinate, int): raise ValueError('Please enter an integer >= 0 for the node_y_coordinate argument.')
    if node_y_coordinate < 0: raise ValueError('Please enter an integer >= 0 for the node_y_coordinate argument.')
    
    # Co-ordinates of left branch is [x, y+1].
    l_value = highest_value_triangle[node_y_coordinate+1][node_x_coordinate]

    # Co-ordinates of right branch is [x+1, y+1].
    r_value = highest_value_triangle[node_y_coordinate+1][node_x_coordinate+1]
    
    # Return highest branch information.
    if l_value > r_value:
        return l_value, node_x_coordinate
    else:
        return r_value, node_x_coordinate + 1

In [4]:
''' 
Searches up through tree, creating the highest value tree and the coordinate tree.

eg:
>>> best_branch(2, 3)
'''

def search_tree() -> None:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Searches up through tree, creating the highest value tree and the coordinate tree.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> best_branch(2, 3)
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    for y in reversed(range(len(triangle)-1)):
        for x in range(len(triangle[y])):
            best_branch_info = best_branch(x, y)
            highest_value_triangle[y][x] += highest_value_triangle[y+1][best_branch_info[1]]
            coordinate_triangle[y][x] = coordinate_triangle[y][x] + coordinate_triangle[y+1][best_branch_info[1]]

In [5]:
search_tree()

In [6]:
solution = highest_value_triangle[0][0]
solution_coordinates = coordinate_triangle[0][0]

solution_chain = []
y = 0
for coordinate in solution_coordinates:
    solution_chain.append(triangle[y][coordinate])
    y+=1

In [7]:
print(f'The maximum total from top to bottom of the triangle is {solution} which is the sum of the numbers:')
print(solution_chain)

The maximum total from top to bottom of the triangle is 1074 which is the sum of the numbers:
[75, 64, 82, 87, 82, 75, 73, 28, 83, 32, 91, 78, 58, 73, 93]


### Problem 18 Solution:

The maximum total from top to bottom of the triangle is 1074 which is the sum of the numbers:

75, 64, 82, 87, 82, 75, 73, 28, 83, 32, 91, 78, 58, 73, 93.

\begin{align}
    &{\color{red}7\color{red}5}\\
    95 &\ {\color{red}6\color{red}4}\\
    17\ &47\ {\color{red}8\color{red}2}\\
    18\ 35 &\ {\color{red}8\color{red}7}\ 10\\
    20\ 04\ &{\color{red}8\color{red}2}\ 47\ 65\\
    19\ 01\ 23 &\ {\color{red}7\color{red}5}\ 03\ 34\\
    88\ 02\ 77\ &{\color{red}7\color{red}3}\ 07\ 63\ 67\\
    99\ 65\ 04\ {\color{red}2\color{red}8} &\ 06\ 16\ 70\ 92\\
    41\ 41\ 26\ 56\ &{\color{red}8\color{red}3}\ 40\ 80\ 70\ 33\\
    41\ 48\ 72\ 33\ 47 &\ {\color{red}3\color{red}2}\ 37\ 16\ 94\ 29\\
    53\ 71\ 44\ 65\ 25\ &43\ {\color{red}9\color{red}1}\ 52\ 97\ 51\ 14\\
    70\ 11\ 33\ 28\ 77\ 73 &\ 17\ {\color{red}7\color{red}8}\ 39\ 68\ 17\ 57\\
    91\ 71\ 52\ 38\ 17\ 14\ &91\ 43\ {\color{red}5\color{red}8}\ 50\ 27\ 29\ 48\\
    63\ 66\ 04\ 68\ 89\ 53\ 67 &\ 30\ {\color{red}7\color{red}3}\ 16\ 69\ 87\ 40\ 31\\
    04\ 62\ 98\ 27\ 23\ 09\ 70\ &98\ 73\ {\color{red}9\color{red}3}\ 38\ 53\ 60\ 04\ 23\\
\end{align}